File: frontend/src/stores/task.ts
import { defineStore } from 'pinia'
import { tasksApi } from '@/api'

interface TaskStep {
  step_id: string
  step_index: number
  action: string
  target: string
  value: string
  status: string
  screenshot_before?: string
  screenshot_after?: string
  retry_count: number
  duration_ms?: number
  error_detail?: string
}

interface Task {
  task_id: string
  instruction: string
  script_id: string
  device_id: string
  status: string
  total_steps: number
  completed_steps: number
  error_type?: string
  error_message?: string
  report_url?: string
  duration_ms?: number
  created_at: string
  steps?: TaskStep[]
}

interface TaskState {
  tasks: Task[]
  currentTask: Task | null
  total: number
  loading: boolean
}

export const useTaskStore = defineStore('task', {
  state: (): TaskState => ({
    tasks: [],
    currentTask: null,
    total: 0,
    loading: false,
  }),

  getters: {
    pendingTasks: (state) => state.tasks.filter((t) => t.status === 'pending'),
    runningTasks: (state) => state.tasks.filter((t) => t.status === 'running'),
    completedTasks: (state) => state.tasks.filter((t) => t.status === 'completed'),
    failedTasks: (state) => state.tasks.filter((t) => t.status === 'failed'),
  },

  actions: {
    async fetchTasks(skip = 0, limit = 20) {
      this.loading = true
      try {
        const response = await tasksApi.list(skip, limit)
        this.tasks = response.data.items
        this.total = response.data.total
      } finally {
        this.loading = false
      }
    },

    async fetchTask(taskId: string) {
      this.loading = true
      try {
        const response = await tasksApi.get(taskId)
        this.currentTask = response.data
        return this.currentTask
      } finally {
        this.loading = false
      }
    },

    async createTask(instruction: string, deviceId?: string) {
      const response = await tasksApi.create({
        instruction,
        device_id: deviceId,
      })
      const task = response.data
      this.tasks.unshift(task)
      return task
    },

    updateTaskInList(taskId: string, updates: Partial<Task>) {
      const index = this.tasks.findIndex((t) => t.task_id === taskId)
      if (index !== -1) {
        this.tasks[index] = { ...this.tasks[index], ...updates }
      }
      if (this.currentTask?.task_id === taskId) {
        this.currentTask = { ...this.currentTask, ...updates }
      }
    },
  },
})

File: backend/app/schemas/task.py
"""Task schemas."""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class TaskStepBase(BaseModel):
    """Base task step schema."""

    action: str = Field(..., max_length=64)
    target: Optional[str] = Field(None, max_length=255)
    value: Optional[str] = None


class TaskStepCreate(TaskStepBase):
    """Schema for creating a task step."""

    step_index: int
    task_id: str


class TaskStepResponse(TaskStepBase):
    """Schema for task step response."""

    step_id: str
    task_id: str
    step_index: int
    status: str
    screenshot_before: Optional[str] = None
    screenshot_after: Optional[str] = None
    retry_count: int
    fix_applied: Optional[str] = None
    error_detail: Optional[str] = None
    duration_ms: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TaskBase(BaseModel):
    """Base task schema."""

    instruction: str


class TaskCreate(TaskBase):
    """Schema for creating a task."""

    device_id: Optional[str] = None


class TaskUpdate(BaseModel):
    """Schema for updating a task."""

    status: Optional[str] = Field(None, max_length=32)
    device_id: Optional[str] = Field(None, max_length=64)
    completed_steps: Optional[int] = None
    error_type: Optional[str] = Field(None, max_length=64)
    error_message: Optional[str] = None
    report_url: Optional[str] = Field(None, max_length=512)
    duration_ms: Optional[int] = None


class TaskResponse(TaskBase):
    """Schema for task response."""

    task_id: str
    user_id: int
    script_id: Optional[str] = None
    device_id: Optional[str] = None
    status: str
    total_steps: int
    completed_steps: int
    error_type: Optional[str] = None
    error_message: Optional[str] = None
    report_url: Optional[str] = None
    duration_ms: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TaskDetailResponse(TaskResponse):
    """Schema for task detail with steps."""

    steps: list[TaskStepResponse] = []


class TaskListResponse(BaseModel):
    """Schema for paginated task list."""

    items: list[TaskResponse]
    total: int
